src/ingest-pipeline/misc/tools/survey.py

```diff
@@ -325,15 +376,164 @@ def all_dataset_uuids(self):
+                samp = self.entity_factory.get(p_uuid)
+                assert isinstance(samp, (Sample, Donor)), 'was expecting a sample or donor?'
+                s_t.add(samp.submission_id)
+            sample_submission_id = str(s_t)
+            sample_hubmap_id = (samp.hubmap_id if len(other_parent_uuids) == 1
+                                  else 'multiple')
+            parent_dataset = None
+        else:
+            sample_submission_id = sample_hubmap_id = None
+            parent_dataset = (self.parent_dataset_uuids[0] if len(self.parent_dataset_uuids) == 1
+                              else 'multiple')
+        return parent_dataset, sample_submission_id, sample_hubmap_id
+
+
+    def build_rec(self, include_all_children=False):
+        """
+        Returns a dict containing:
+        
+        uuid
+        hubmap_id
+        status
+        group_name
+        note 
+
+        """
+        rec = {'uuid': self.uuid, 'hubmap_id': self.hubmap_id, 'status': self.status,
+               'group_name': self.group_name, 'note': ''}
+        rec['note'] = '; '.join([rec['note']] + self.notes)
+        return rec
+            
+    def all_uuids(self):
+        """
+        Returns a list of unique UUIDs associated with this entity
+        """
+        return super().all_uuids() + self.kid_uuids + self.parent_uuids
+
+    def all_dataset_uuids(self):
+        """
+        Returns a list of unique dataset or support UUIDs associated with this entity
+        """
+        return super().all_dataset_uuids() + self.kid_dataset_uuids + self.parent_dataset_uuids
+    
+
 class Sample(Entity):
     def __init__(self, prop_dct, entity_factory):
         super().__init__(prop_dct, entity_factory)
         assert prop_dct['entity_type'] == 'Sample', f"uuid {uuid} is a {prop_dct['entity_type']}"
         self.hubmap_id = prop_dct['hubmap_id']
-        self.donor_hubmap_id = prop_dct['donor']['hubmap_id']
+        self._donor_hubmap_id = None
         self.submission_id = prop_dct['submission_id']
-        self.donor_submission_id = prop_dct['donor']['submission_id']
-        self.donor_uuid = prop_dct['donor']['uuid']
+        self._donor_submission_id = None
+        self._donor_uuid = None
+        assert 'direct_ancestor' in prop_dct, 'prop_dct for Sample with no direct_ancestors'
+        direct_ancestors_list = [prop_dct['direct_ancestor']]
+        self.parent_uuids = [elt['uuid'] for elt in direct_ancestors_list]
+        self._organ = prop_dct.get('organ', None)
+
+    @property
+    def organ(self):
+        if self._organ is None:
+            # There should only be one parent, another Sample
+            for parent_uuid in self.parent_uuids:
+                parent_entity = self.entity_factory.get(parent_uuid)
+                assert isinstance(parent_entity, Sample), 'Outermost sample has no organ code?'
+                self._organ = parent_entity.organ
+        return self._organ
+
+    @property
+    def donor_uuid(self):
```

Comment:
Won't this always break on the first parent_uuid in the list?

Reply:
Yeah, I agree :-/ .  I'll restructure that.